fsmonitor.c

```diff
@@ -56,11 +57,13 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
 
 void fill_fsmonitor_bitmap(struct index_state *istate)
 {
-	unsigned int i;
+	unsigned int i, removed = 0;
 	istate->fsmonitor_dirty = ewah_new();
 	for (i = 0; i < istate->cache_nr; i++)
-		if (!(istate->cache[i]->ce_flags & CE_FSMONITOR_VALID))
-			ewah_set(istate->fsmonitor_dirty, i);
+		if (istate->cache[i]->ce_flags & CE_REMOVE)
+			removed++;
+		else if (!(istate->cache[i]->ce_flags & CE_FSMONITOR_VALID))
+			ewah_set(istate->fsmonitor_dirty, i - removed);
```

Comment:
Assuming the change looks good and passes existing tests, are there new tests we can add?

Reply:
Absolutely! The best place would be [t/t7519-status-fsmonitor.sh](https://github.com/microsoft/git/blob/vfs-2.23.0/t/t7519-status-fsmonitor.sh), I think.

So I played around with this a bit, and it seems as if the `git restore` call (or the `git checkout` call, if you're old school) is crucial for this bug to reproduce. This patch seems to work for me: reverting your change to `fsmonitor.c` causes the test case to fail due to a segmentation fault, reinstating the changes fixes it again:

```diff
diff --git a/fsmonitor.c b/fsmonitor.c
index 231e83a94db..0119eda37ce 100644
--- a/fsmonitor.c
+++ b/fsmonitor.c
@@ -56,11 +56,15 @@ int read_fsmonitor_extension(struct index_state *istate, const void *data,
 
 void fill_fsmonitor_bitmap(struct index_state *istate)
 {
-	unsigned int i;
+	unsigned int i, removed = 0;
 	istate->fsmonitor_dirty = ewah_new();
-	for (i = 0; i < istate->cache_nr; i++)
-		if (!(istate->cache[i]->ce_flags & CE_FSMONITOR_VALID))
-			ewah_set(istate->fsmonitor_dirty, i);
+	for (i = 0; i < istate->cache_nr; i++) {
+		unsigned int flags = istate->cache[i]->ce_flags;
+		if (flags & CE_REMOVE)
+			removed++;
+		else if (!(flags & CE_FSMONITOR_VALID))
+			ewah_set(istate->fsmonitor_dirty, i - removed);
+	}
 }
 
 void write_fsmonitor_extension(struct strbuf *sb, struct index_state *istate)
diff --git a/t/t7519-status-fsmonitor.sh b/t/t7519-status-fsmonitor.sh
index 81a375fa0ff..9e9621a3548 100755
--- a/t/t7519-status-fsmonitor.sh
+++ b/t/t7519-status-fsmonitor.sh
@@ -354,4 +354,15 @@ test_expect_success 'discard_index() also discards fsmonitor info' '
 	test_cmp expect actual
 '
 
+test_expect_success 'fsmonitor extension ignores removed index entries' '
+	test_commit initial &&
+	removed=$(seq 1 100 | sed "s/^/z/") &&
+	touch $removed &&
+	git add $removed &&
+	test_config core.fsmonitor "$TEST_DIRECTORY/t7519/fsmonitor-env" &&
+	FSMONITOR_LIST="$removed" git restore -S $removed &&
+	FSMONITOR_LIST="$removed" git status
+'
+
 test_done
diff --git a/t/t7519/fsmonitor-env b/t/t7519/fsmonitor-env
new file mode 100755
index 00000000000..8f1f7ab164e
--- /dev/null
+++ b/t/t7519/fsmonitor-env
@@ -0,0 +1,24 @@
+#!/bin/sh
+#
+# An test hook script to integrate with git to test fsmonitor.
+#
+# The hook is passed a version (currently 1) and a time in nanoseconds
+# formatted as a string and outputs to stdout all files that have been
+# modified since the given time. Paths must be relative to the root of
+# the working tree and separated by a single NUL.
+#
+#echo "$0 $*" >&2
+
+if test "$#" -ne 2
+then
+	echo "$0: exactly 2 arguments expected" >&2
+	exit 2
+fi
+
+if test "$1" != 1
+then
+	echo "Unsupported core.fsmonitor hook version." >&2
+	exit 1
+fi
+
+printf '%s\n' $FSMONITOR_LIST
```

Reply:
Sorry if this is a dumb question (ipad) but are there any places where it would be helpful to assert the length of the bitmap and the cache entry array are equal?  Just trying to verify that any code that manipulates one also updates the other.

Reply:
> are there any places where it would be helpful to assert the length of the bitmap and the cache entry array are equal? Just trying to verify that any code that manipulates one also updates the other.

@dscho please correct me if my understanding is wrong.  

Based on my reading of the code the bitmap length should always be <= to the length of the cache entry array, but it won't necessarily be equal to the length of the cache entries.  `ewah_set` grows the bitmap only when needed, and so if the last X cache entries have the `CE_FSMONITOR_VALID` flag set, the length of `fsmonitor_dirty` will be `istate->cache_nr - X`.

I will see if there is anywhere else that `fsmonitor_dirty` is being adjusted where an `assert` that its length is less than `istate->cache_nr` is appropriate.

Reply:
> I will see if there is anywhere else that fsmonitor_dirty is being adjusted where an assert that its length is less than istate->cache_nr is appropriate.

I've found a few spots (maybe too many?) and will include them in my next push, please let me know what you think.

Reply:
@dscho thanks for writing up a test!  For my own understanding, is this:

`sed "s/^/z/"`

To ensure the entries being added/removed appear at the end of the cache?

Reply:
Yes.